pytorch_lightning/loops/epoch/training_epoch_loop.py

```diff
@@ -98,7 +98,8 @@ def reset(self) -> None:
         # track epoch output
         self._epoch_output = [[] for _ in range(self.batch_loop.num_active_optimizers(self.total_batch_idx))]
 
-        if not self.restarting:
+        ended = self._num_training_batches_reached()
+        if not self.restarting or ended:
```

Comment:
```suggestion
        if not self.restarting or self._num_training_batches_reached():
```

nit: no need for a separate variable if only used once :)